andela/jubilee-backend
# #168243791 Integrate User Roles to Signup features with permissions

#### What does this PR do?
Integrates the User Roles features with the signup, It also checks user permissions before accessing specific routes, and integrates the signup token functionality

#### Description of Task to be completed?
- generate unique signup token for company and suppliers
- Refactor middleware to decipher tokens upon user signup
- Assign Default roles upon signup for all users
- create permissions class for backend permission checks on all routes

#### How should this be manually tested?
- Clone the repo and cd into app root
-  Setup necessary environment variables for testing environment (check .sample.env)
- Run docker-compose up --build to build docker image or run locally using `npm run start:dev`
- Run `npm run pretest` to migrate and seed database 
-  run `npm run test` to run all tests
- On postman, run `localhost:3000/api/auth/signup/supplier` with user json parameters to signup a supplier:
```
{
	"firstName": "chike",
	"lastName": "Nonso",
	"email": "chike@admin.com",
	"password": "tmobnvar..Ds66u",
	"phoneNumber": "(+234) 090-3472-3889",
	"companyName": "Jubilee",
	"companyAddress": "23 elmo plaza, off rainoil road, lekki",
	"categoryOfServiceId": 1
}
```
- or on route ``localhost:3000/api/auth/signup/company`, user json parameters to signup a company
```
{
	"firstName": "Tundev",
	"lastName": "Namib",
	"email": "tundev@admin.com",
	"password": "tmobnvar..Ds66u",
	"companyName": "Andelaa",
	"companyAddress": "23 elmo plaza, off rainoil road, lekki",
	"companySizeId": 1,
	"companyPlanId": 1
}
```
- In postman, on route `localhost:3000/api/auth/signup/user` use the returned companyToken to signup a new user:
```
{
	"firstName": "Davis",
	"lastName": "Okra",
	"email": "Davis@okra.com",
	"password": "tmobnvarq.ss66u",
	"companyName": "Jubilee",
	"signupToken": "[PLACE GENERATED TOKEN HERE]"
}
```

#### Any background context you want to provide?
none

### screenshots (if appropriate)

<img width="1176" alt="Screenshot 2019-09-05 at 4 21 00 AM" src="https://user-images.githubusercontent.com/37340699/64324399-3adfeb80-cfbe-11e9-8545-04256efa89ff.png">
<img width="1176" alt="Screenshot 2019-09-05 at 5 17 00 AM" src="https://user-images.githubusercontent.com/37340699/64324401-3b788200-cfbe-11e9-83ac-1b1ff7053975.png">

#### What are the relevant pivotal tracker stories?
[#168243791](https://www.pivotaltracker.com/story/show/168243791)
